[Sean Lam] iP

LJXSean opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- SUPER FAST ~FAST~ to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks 🎉 
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

TheStryver committed:
Level-1

TheStryver committed:
Level-2

TheStryver committed:
Level-3

TheStryver committed:
Level-4

TheStryver committed:
Level-5

TheStryver committed:
Level-6

TheStryver committed:
TextUiTesting

TheStryver committed:
Implement switch statements instead of if-else

TheStryver committed:
Add save functionality

TheStryver committed:
Add dateTime recognition

TheStryver committed:
Merge branch 'branch-Level-7'

TheStryver committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

TheStryver committed:
Make the code follow OOP

TheStryver committed:
Grouped relevant files into duke package

TheStryver committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

TheStryver committed:
Add Junit tests

TheStryver committed:
Add JavaDocs

TheStryver committed:
Add find functionality

TheStryver committed:
Merge branch 'branch-Level-9'

TheStryver committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/TaskList.java

Polygonalr commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,111 @@
+                            input = sc.nextLine();
+                        } else {
+                            throw new DukeException("OOPS!!! The description of a todo cannot be empty.");
+                        }
+                        break;
+                    case "deadline":
+                        if (!Parser.isValidDescription(input)) {
+                            throw new DukeException("OOPS!!! The description of a duke.Deadline cannot be empty.");
+                        } else if (!Parser.isValidDateTime(input)) {
+                            throw new DukeException("OOPS!!! The time and date of the duke.Deadline cannot be empty.");
+                        } else {
+                            //eg. by 2019-10-03 18:00
+                            Task toAdd = new Deadline(Parser.getDescription(input),
+                                    Parser.getDate(input),
+                                    Parser.getFrom(input));
+                            itemList.addTask(toAdd);
+                            input = sc.nextLine();
+                        }
+                        break;
+                    case "event":
+                        if (!Parser.isValidDescription(input)) {
+                            throw new DukeException("OOPS!!! The description of a duke.Event cannot be empty.");
+                        } else if (!Parser.isValidDateTime(input)) {
+                            throw new DukeException("OOPS!!! The time and date of the duke.Event cannot be empty.");
+                        } else {
+                            Task toAdd = new Event(Parser.getDescription(input),
+                                    Parser.getDate(input),
+                                    Parser.getFrom(input),
+                                    Parser.getTo(input));
+                            itemList.addTask(toAdd);
+                            input = sc.nextLine();
+                        }
+                        break;
+                    case "find":
+                        itemList.findTask(Parser.getDescription(input));
+                        input = sc.nextLine();
+                        break;
+                    // unrecognised commands
+                    default:
+                        throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
```
No indentation for `case` clauses according to the module's [Java Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html).

LJXSean replied:
Ok thanks for the advice

Polygonalr commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,18 @@
+public class DeadlineTest {
+    Deadline testEvent = new Deadline("lunch", "2022-02-03", "12:00");
+    @Test
+    public void testToString(){
+        assertEquals(testEvent.toString(), "[D][ ] lunch (by: Feb 3 2022, 1200PM)");
+    }
+
+    @Test
+    public void testToFile(){
```
Remember your spaces before the `{` in your test files.

Polygonalr commented on src/test/java/duke/EventTest.java:
```diff
@@ -0,0 +1,18 @@
+public class EventTest {
+    Event testEvent = new Event("lunch", "2022-02-03", "12:00", "14:00");
+    @Test
+    public void testToString(){
+        assertEquals(testEvent.toString(), "[E][ ] lunch (at: Feb 3 2022, 1200PM to 0200PM)");
+    }
+
+    @Test
+    public void testToFile(){
```
Same as `DeadlineTest.java`.

Polygonalr commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,111 @@
+        System.out.print("Hello! I'm duke.Duke \nWhat can I do for you? \n");
+
+        String input = sc.nextLine();
+        boolean carryOn = true;
```
Do you think this boolean can be named more appropriately here (i.e. `canCarryOn`)?

Polygonalr reviewed (commented):
LGTM aside from some coding conventions!

ElginL commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+     * @param date Date of deadline
+     * @param time Time of deadline
+     */
+    Deadline(String description, String date, String time) {
```
I think there might be a missing access modifier right here. It would probably be better to have an access modifier to avoid vagueness

ElginL commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,58 @@
+    public String toString() {
+        return type + super.toString() + " (at: " + date.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                + ", " + from.format(DateTimeFormatter.ofPattern("hhmma")) + " to "
+                + to.format(DateTimeFormatter.ofPattern("hhmma")) + ")";
```
I didn't like this because it is a complicated expression and it took quite a bit of effort to figure out what it prints. It is quite prevalent around. Perhaps, I think it would be better if intermediate values are calculated before printing.

ElginL commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,133 @@
+            case "E":
+                toAdd = new Event(details[2], details[3], details[4], details[5]);
+                break;
+        }
```
I think there might be some styling mistakes right here for switch case. According to the styling convention, case should not be indented

ElginL commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,50 @@
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
```
I like how you used verbs for all methods in your classes. Good job on that 👍

ElginL commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,111 @@
+        this.itemList = itemList;
+    }
+
+    public void getUserInputs() {
```
I think this method length is too long (violates the recommended "Avoid Long Methods" >=30LOC ).
Perhaps an improvement might be to abstract out the logic within each case.

ElginL commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,111 @@
+                            input = sc.nextLine();
+                        } else {
+                            throw new DukeException("OOPS!!! The description of a todo cannot be empty.");
+                        }
```
This code works, but I think there could be a little room for improvement. The Happy Path could've been made more prominent.

```java
if (!Parser.isValidDescription(input)) {
  throw new DukeException("OOPS!!! The description of a todo cannot be empty.");
}

Task toAdd = new ToDo(Parser.getDescription(input));
itemList.addTask(toAdd);
input = sc.nextLine();
```

ElginL reviewed (commented):
Good work on using OOP, mostly good and works. However, there could still be a little room of improvement in terms of code quality. 😃

TheStryver committed:
Add GUI

TheStryver committed:
Edits

TheStryver committed:
Merge branch 'branch-Level-10'

TheStryver committed:
Add assertions

TheStryver committed:
Fix code quality violations

Priority when writing code is on functionality and making it work. There are several areas where the code quality is subpar.

Code quality has to be improved to increase readability and make it easier to debug hidden bugs.

Let's improve code quality by closely following the code quality standards specified.

LJXSean committed:
Merge pull request #1 from LJXSean/branch-A-codeQuality

Fix code quality violations

LJXSean committed:
Merge branch 'master' into branch-A-Assertions

LJXSean committed:
Merge pull request #2 from LJXSean/branch-A-Assertions

Add assertions

TheStryver committed:
Merge branch 'master' of https://github.com/LJXSean/ip

TheStryver committed:
Merge branch 'branch-A-Assertions' of https://github.com/LJXSean/ip into branch-A-Assertions

TheStryver committed:
Merge branch 'master' into branch-A-Assertions

TheStryver committed:
Add help command

There are many commands that can be used and each command requires the
user to follow the specified input format precisely.

New users to the app need to be able to find out what commands are available
and the specific ways to the details need to be typed in. Regular users might
also forget certain commands. Hence a help command is required.

Let's add a help command to list out the possible inputs and their required input
formats.

This allows users to quickly refer to it when needed while using Duke.

LJXSean committed:
Merge pull request #3 from LJXSean/branch-C-Help

Add help command

TheStryver committed:
Merge branch 'master' of https://github.com/LJXSean/ip

TheStryver committed:
Update readMe

TheStryver committed:
Update formatting in readMe

TheStryver committed:
Update formatting in readMe
